sdk/eidos_sdk/system/agent_controller.py

```diff
@@ -180,18 +188,24 @@ async def send_response(self, handler: EidosHandler, process: ProcessDoc, **kwar
 
         process.state = state_change_event.state
         if final_event.is_root_and_type(ErrorEvent):
-            process.data = final_event.reason
+            process.error_info = final_event.reason
+            return self.doc_to_response(process, None)
         else:
             if result_object:
-                process.data = result_object
+                data = result_object
             else:
-                process.data = string_result
-        return self.doc_to_response(process)
+                data = string_result
+            return self.doc_to_response(process, data)
 
-    def agent_event_stream(self, handler, process, **kwargs) -> AsyncIterator[StreamEvent]:
+    async def agent_event_stream(self, handler, process, **kwargs) -> AsyncIterator[StreamEvent]:
         is_async_gen = inspect.isasyncgenfunction(handler.fn)
         stream = handler.fn(self.agent, **kwargs) if is_async_gen else self.stream_agent_fn(handler, **kwargs)
-        return self.stream_agent_iterator(stream, process, handler.name)
+        events_to_store = []
+        async for event in self.stream_agent_iterator(stream, process, handler.name):
+            events_to_store.append(event)
+            yield event
+
+        await store_events(self.name, process.record_id, events_to_store)
```

Comment:
you want to store the events in self.agent_event_stream

send_response is just for application/json response type

Reply:
That's where to code is..
    async def agent_event_stream(self, handler, process, **kwargs) -> AsyncIterator[StreamEvent]:
        is_async_gen = inspect.isasyncgenfunction(handler.fn)
        stream = handler.fn(self.agent, **kwargs) if is_async_gen else self.stream_agent_fn(handler, **kwargs)
        events_to_store = []
        async for event in self.stream_agent_iterator(stream, process, handler.name):
            events_to_store.append(event)
            yield event

        await store_events(self.name, process.record_id, events_to_store)